src/Servers/Kestrel/Core/src/KestrelServer.cs

```diff
@@ -129,54 +144,52 @@ public async Task StartAsync<TContext>(IHttpApplication<TContext> application, C
 
                 async Task OnBind(ListenOptions options)
                 {
-                    // Add the HTTP middleware as the terminal connection middleware
-                    options.UseHttpServer(ServiceContext, application, options.Protocols);
-
-                    var connectionDelegate = options.Build();
-
-                    // Add the connection limit middleware
-                    if (Options.Limits.MaxConcurrentConnections.HasValue)
-                    {
-                        connectionDelegate = new ConnectionLimitMiddleware(connectionDelegate, Options.Limits.MaxConcurrentConnections.Value, Trace).OnConnectionAsync;
-                    }
-
-                    var connectionDispatcher = new ConnectionDispatcher(ServiceContext, connectionDelegate);
-
-                    IConnectionListenerFactory factory = null;
-                    if (options.Protocols >= HttpProtocols.Http3)
+                    // INVESTIGATE: For some reason, MsQuic needs to bind before
+                    // sockets for it to successfully listen. It also seems racy.
+                    if ((options.Protocols & HttpProtocols.Http3) == HttpProtocols.Http3)
                     {
-                        foreach (var transportFactory in _transportFactories)
+                        if (_multiplexedTransportFactories == null)
```

Comment:
DI gives an empty IEnumerable if nothing is registered. I'd probably add a test for this case.

```suggestion
                        if (_multiplexedTransportFactories == null || _multiplexedTransportFactories.Count == 0)
```